beater/request/result.go

```diff
@@ -160,20 +120,128 @@ type Result struct {
 	Stacktrace string
 }
 
-//SendStatus initiates response writing for a given context
-//TODO: move to Context when reworking response handling.
-func SendStatus(c *Context, res Result) {
-	if res.Err != nil {
-		body := map[string]interface{}{"error": res.Err.Error()}
-		//TODO: refactor response handling: get rid of additional `error` and just pass in error
-		c.SendError(body, body, res.StatusCode)
+// Reset sets result to it's empty values
+func (r *Result) Reset() {
+	r.ID = IDUnset
+	r.StatusCode = http.StatusOK
+	r.Keyword = ""
+	r.Body = nil
+	r.Err = nil
+	r.Stacktrace = ""
+}
+
+// SetDefault derives information about the result solely from the ID.
+func (r *Result) SetDefault(id ResultID) {
+	r.set(id, nil, nil)
+}
+
+// SetWithError derives information about the result from the given ID and the error.
+// The body is derived from the error in case the result describes a failure.
+func (r *Result) SetWithError(id ResultID, err error) {
+	r.set(id, nil, err)
+}
+
+// SetWithBody derives information about the result from the given ID. The body is set to the passed value.
+func (r *Result) SetWithBody(id ResultID, body interface{}) {
+	r.set(id, body, nil)
+}
+
+// Set allows for the most flexibility in setting a result's properties.
+// The error and body information are derived from the given parameters.
+func (r *Result) Set(id ResultID, statusCode int, keyword string, body interface{}, err error) {
+	if r == nil {
 		return
+	}
+	r.ID = id
+	r.StatusCode = statusCode
+	r.Keyword = keyword
+	r.Body = body
+	r.Err = err
 
+	if r.Failure() {
+		if err == nil {
+			r.Err = errors.New(keyword)
+		}
+		if r.Body == nil {
+			r.Body = r.Err.Error()
+		}
 	}
-	if res.Body == nil {
-		c.WriteHeader(res.StatusCode)
+}
+
+// Failure returns a bool indicating whether it is describing a successful result or not
+func (r *Result) Failure() bool {
+	return r.StatusCode >= http.StatusBadRequest
+}
+
+func (r *Result) set(id ResultID, body interface{}, err error) {
+	if r == nil {
 		return
 	}
+	statusCode := http.StatusInternalServerError
+	keyword := KeywordResponseErrorsInternal
+
+	switch id {
```

Comment:
instead of having separate constants for everything and manually matching them here, why not have a map of `<ResultId, Pair(code, keyword)>` and replace the whole `switch` with a map lookup?

Reply:
You are right, I'll change that.